flow/ActorLineageSet.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * ActorLineageSet.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "flow/flow.h"
+#include <boost/lockfree/queue.hpp>
+
+class ActorLineageSet {
+public:
+	// The type we use for lookup into the set. Gets assigned during insert
+	using Index = unsigned;
+	// For now we use a fixed size capacity
+	constexpr static Index CAPACITY = 1024;
+	constexpr static Index npos = std::numeric_limits<Index>::max();
+
+	explicit ActorLineageSet();
+	ActorLineageSet(const ActorLineageSet&) = delete;
+	ActorLineageSet& operator=(const ActorLineageSet&) = delete;
+
+	// Returns the number of elements at the time of calling. Keep in mind that this is a lockfree data structure, so
+	// the actual size might change anytime after or even during the call. This function only guarantees that the size
+	// was whatever the method returns at one point between the start and the end of the function call. The safest way
+	// to handle this is by assuming that this returns an estimate.
+	unsigned size();
+
+	Index insert(const Reference<ActorLineage>& lineage);
+	void erase(Index idx);
+	std::vector<Reference<ActorLineage>> copy();
+
+private:
+	static constexpr uintptr_t FREE = 0b1;
+	static constexpr uintptr_t LOCK = 0b10;
+	std::atomic<unsigned> _size = 0;
+	std::vector<std::atomic<std::uintptr_t>> _set;
+	boost::lockfree::queue<Index, boost::lockfree::fixed_sized<true>, boost::lockfree::capacity<CAPACITY>> freeQueue;
+	boost::lockfree::queue<ActorLineage*, boost::lockfree::fixed_sized<true>, boost::lockfree::capacity<CAPACITY>>
+	    freeList;
+};
+
+ActorLineageSet::ActorLineageSet() {
+	// insert the free indexes in reverse order
+	for (unsigned i = CAPACITY; i > 0; --i) {
+		freeQueue.push(i - 1);
+		_set[i] = uintptr_t(1);
```

Comment:
Doesn't `_set` need to be initialized with some capacity before accessing an arbitrary index?

Reply:
yes you're right. Or maybe we should even use a `std::array` instead.

Reply:
If capacity is fixed I think that makes sense.